mlflow/llama_index/tracer.py

```diff
@@ -293,3 +340,99 @@ def _extract_token_usage(
                     if (v := additional_kwargs.get(k)) is not None:
                         usage[k] = v
             return usage
+
+
+_StreamEndEvents = Union[LLMChatEndEvent, LLMCompletionEndEvent, ExceptionEvent]
+
+
+class StreamResolver:
+    """
+    A class is responsible for closing the pending streaming spans that are waiting
+    for the stream to be exhausted. Once the associated stream is exhausted, this
+    class will resolve the span, as well as recursively resolve the parent spans
+    that returns the same (or derived) stream.
+    """
+
+    def __init__(self, client: mlflow.MlflowClient):
+        self._client = client
+        self._span_id_to_span_and_gen: Dict[str, Tuple[LiveSpan, Generator]] = {}
+
+    def is_streaming_result(self, result: Any) -> bool:
+        return (
+            inspect.isgenerator(result)  # noqa: SIM101
+            or isinstance(result, (StreamingResponse, AsyncStreamingResponse))
+            or isinstance(result, StreamingAgentChatResponse)
+            or (isinstance(result, TaskStepOutput) and self.is_streaming_result(result.output))
+        )
+
+    def register_stream_span(self, span: LiveSpan, result: Any) -> bool:
+        """
+        Register the pending streaming span with the associated generator.
+
+        Args:
+            span: The span that has a streaming output.
+            result: The streaming result that is being processed.
+
+        Returns:
+            True if the span is registered successfully, False otherwise.
+        """
+        if inspect.isgenerator(result):
+            stream = result
+        elif isinstance(result, (StreamingResponse, AsyncStreamingResponse)):
+            stream = result.response_gen
+        elif isinstance(result, StreamingAgentChatResponse):
+            stream = result.chat_stream
+        elif isinstance(result, TaskStepOutput):
+            stream = result.output.chat_stream
+        else:
+            _logger.warning(
+                f"Unsupported streaming response type: {type(result)}. The span will "
+                "be ended immediately without waiting for the stream to finish."
+            )
+            return False
+
+        if inspect.getgeneratorstate(stream) == inspect.GEN_CLOSED:
+            # Not registering the span because the generator is already exhausted
+            return False
+
+        self._span_id_to_span_and_gen[span.span_id] = (span, stream)
+        return True
+
+    def resolve(self, span: LiveSpan, event: _StreamEndEvents):
+        """
+        Finish the streaming span and recursively resolve the parent spans that
+        returns the same (or derived) stream.
+        """
+        _, stream = self._span_id_to_span_and_gen.pop(span.span_id, (None, None))
+        if not stream:
+            return
```

Comment:
Should we differentiate between `span.span_id is not in the dict` versus `stream == None`?

Reply:
Conceptually yes, but practically the action we take is same - just skip handling the parent.